src/NATS.Client.Core/Commands/PublishCommand.cs

```diff
@@ -32,12 +35,17 @@ public static PublishCommand<T> Create(ObjectPool pool, string subject, string?
         result._serializer = serializer;
         result._cancellationToken = cancellationToken;
 
+        // Serialize data as soon as possible to propagate any exceptions
+        // to the caller so that publish method will throw the exception
+        result._buffer.Reset();
+        serializer.Serialize(result._buffer, value);
+
         return result;
     }
 
     public override void Write(ProtocolWriter writer)
     {
-        writer.WritePublish(_subject!, _replyTo, _headers, _value, _serializer!);
+        writer.WritePublish(_subject!, _replyTo, _headers, new ReadOnlySequence<byte>(_buffer.WrittenMemory));
```

Comment:
With this change, `_serializer` and  `_value` aren't needed anymore and `PublishCommand` doesn't need to be generic.

Reply:
Also, `_buffer` or rather it's underlying array should be freed here. Otherwise, we would keep a potentially huge buffer allocated after sending a large message.

Reply:
>  It also enables serialization work to happen concurrently (for multiple publish operations) instead of serializing all serialization work inside WriteLoopAsync

Good point.

> Also, _buffer or rather it's underlying array should be freed here.

The idea of keeping the _buffer around to avoid GC since it'd be pooled with the command object.

On the other hand we could leave out the buffer management to end developer and just accept ReadOnlySequence or Memory or something or even an IMemoryOwner?

Reply:
Ah, @jasper-d I think I misunderstand your freeing buffer comment (wasn't looking at where in code you commented). You have a point. I thought about this as well. We could resize the underlying array to be smaller or enforce a size limit as it would be by the server anyway.